### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by natefaubion**
This PR improves the performance across the board by better optimizing the sharing on pattern matches. There were a lot of duplicate checks for things like `==` and `<=` in guards. In my tests, about a 2x speedup overall. Ideally, a smarter pattern matcher could perform these optimizations, but for now we can do it manually.

lookup
insert
delete

### CODE REVIEW

**Comment by paf31** [comment]
:+1: Fantastic, thanks!

**Comment by jdegoes** [comment]
:+1:

File: src/Data/Map.purs [hunk 1]
```diff
@@ -116,14 +116,24 @@ checkValid tree = length (nub (allHeights tree)) == one
...
-lookup k (Two left k1 _ _) | k < k1 = lookup k left
-lookup k (Two _ _ _ right) = lookup k right
-lookup k (Three _ k1 v1 _ _ _ _) | k == k1 = Just v1
-lookup k (Three _ _ _ _ k2 v2 _) | k == k2 = Just v2
-lookup k (Three left k1 _ _ _ _ _) | k < k1 = lookup k left
-lookup k (Three _ k1 _ mid k2 _ _) | k1 < k && k <= k2 = lookup k mid
-lookup k (Three _ _ _ _ _ _ right) = lookup k right
+lookup k tree =
+  let comp :: k -> k -> Ordering
+      comp = compare
```
Comment (garyb) [inline src/Data/Map.purs]: I wasn't immediately sure why you did this at first, but I get it. Nice. I'll definitely do some work on this one day :wink:

**Comment by ethul** [comment]
I just tried the 0.5.5 release and I am getting the following on psc 0.7.6.1:

```
Error found:
at /.../bower_components/purescript-maps/src/Data/Map.purs line 132, column 18 - line 132, column 18

  Unable to parse module:
  unexpected ,
  expecting ., {, number, char literal, string literal, "true", "false", [, _, "\\", qualifier, proper name, identifier, (, "case", "if", "do", "let", ::, operator or "of"
```

Was this a breaking change from 0.5.4?

**Comment by garyb** [comment]
Argh, it wasn't supposed to be. Looks like it's the multi-valued case, I thought that went in before 0.8.

@natefaubion perhaps we should use a record/array literal there for now, and then remove again when we make the release for psc 0.8?

**Comment by garyb** [comment]
I've just reverted it for now, I did a version using records and arrays but then was greeted by a bunch of partial pattern warnings. Perhaps there is something we can do pre-0.8 still though.

**Comment by natefaubion** [comment]
> I thought that went in before 0.8.

I would have sworn that as well.

I think records should work (you're just doing useless allocations :/). Arrays will always give you the partial warnings. You could also use other functions, but that may be slower since you can't use `mkFn`, so you'd have the overhead of more currying.

**Comment by natefaubion** [comment]
Both `lookup` and `insert` can be rewritten with single case matching with slightly more verbose branching, but `delete` cannot. Swapping out the multi-case for objects actually makes it about twice as *slow* as the current `delete`. There are fewer perf wins for `delete`, so if we want to make these perf changes compatible for 0.7.6, then should probably just leave `delete` as is.

Do we want to update `lookup` and `insert` as a 0.5.x change, and then fix warnings + `delete` for a 0.6 change?

**Comment by garyb** [comment]
Sounds good to me, thanks for looking into this again.

**Pull request merged**